src/Data/LLVM/BitCode/GetBits.hs

```diff
@@ -82,96 +120,120 @@ instance MonadPlus GetBits where
   mplus = (<|>)
 
 
--- Sub-word Bit Parsing State --------------------------------------------------
-
--- This assumes that we'll pad any incoming bitcode to a 32-bit boundary, but
--- given the use of 32-bit padding already, it seems likely that all bitcode
--- files get padded to a 32-bit boundary.
-data SubWord
-  = SubWord !NumBits !Word32
-    deriving (Show)
-
-aligned :: SubWord
-aligned = SubWord (Bits' 0) 0
-
-splitWord :: NumBits -> SubWord -> (BitString, Either NumBits SubWord)
-splitWord n (SubWord l w)
-  | n <= l = ( toBitString n (fromIntegral w)
-             , Right (SubWord (subtractBitCounts l n) (w `shiftR` bitCount n))
-             )
-  | otherwise = (toBitString l (fromIntegral w), Left (subtractBitCounts n l))
-
--- | @getBitString n@ grabs a @BitString@ of length @n@ from the next incoming
--- word, yielding the remainder partial word.  On @n@ = @0@, it does not
--- actually consume the next word.  Should not be called to read more than one
--- 32-bit word at a time (will fail on @n@ > 32).
-getBitString :: NumBits -> BG.Get (BitString, SubWord)
-getBitString (Bits' 0) = return (mempty, aligned)
-getBitString n | n > Bits' 32 =
-  fail $ "getBitString: refusing to read " ++ show n ++ " (> 32) bits."
-getBitString n = getBitStringPartial n . SubWord (Bits' 32) =<< BG.getWord32le
-
--- | @getBitStringPartial n sw@ returns a @BitString@ of length @n@ from either
--- the current subword @sw@ (with some @l@ bits available), potentially also
--- reading the next incoming word if @n@ > @l@.  Should not be called to read
--- more than one 32-bit word at a time (will fail on @n@ > 32).
-getBitStringPartial :: NumBits -> SubWord -> BG.Get (BitString, SubWord)
-getBitStringPartial n _ | n > Bits' 32 =
-  fail $ "getBitStringPartial: refusing to read " ++ show n ++ " (> 32) bits."
-getBitStringPartial n sw = case splitWord n sw of
-  (bs, Right off) -> return (bs, off)
-  (bs, Left n') -> do
-    (rest, off) <- getBitString n'
-    return (bs `mappend` rest, off)
-
--- | Skip a byte of input, which must be zero.
-skipZeroByte :: BG.Get ()
-skipZeroByte = do
-  x <- BG.getWord8
-  when (x /= 0) $ fail "alignment padding was not zeros"
-
--- | Get a @ByteString@ of @n@ bytes, and then align to 32 bits.
-getByteString :: NumBytes -> BG.Get (ByteString, SubWord)
-getByteString (Bytes' n) = do
-  bs <- BG.getByteString n
-  replicateM_ ((- n) `mod` 4) skipZeroByte
-  return (bs, aligned)
+-- | Extracts an Integer value of the specified number of bits from a ByteString,
+-- starting at the indicated bit position (fails with Left if the range to
+-- extract is not valid... i.e. > bitLimit).  Returns the Integer value along
+-- with the bit position following the extraction.
+
+extractFromByteString :: NumBits -> NumBits -> NumBits -> ByteString
```

Comment:
Worth including some Haddock comments here to distinguish all of the `NumBits` arguments.